---
title: "Class12 Homework "
author: "Jordan Prych A17080226"
format: pdf
toc: TRUE
---
```{r}
knitr::opts_chunk$set(echo=TRUE)
```


## Section 1. Proportiona of G/G in a Population

Downloaded CSV file from Ensemble

Here we read this CSV file to determine allele frequency 

```{r}
MXL <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378 (1).csv")
head(MXL)
```
```{r}
MXL$Genotype..forward.strand.
```

```{r}
table(MXL$Genotype..forward.strand.)
```

```{r}
table(MXL$Genotype..forward.strand.)/nrow(MXL)
```

## Section 4. Population Scale Analysis Homework 

One sample is obviously not enough to know what is happening in a population. You are interested in assessing genetic differences on a population scale. 

>Q13: Read this file into R and determine the sample size for each genotype and their 
corresponding median expression levels for each of these genotypes.

How many samples do we have? 

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
nrow(expr)
```
There are 462 individuals(this is the sample size). 

```{r}
table(expr$geno)
```
Let's find the median expression levels for each genotype from the boxplot below. 

```{r}
median <- tapply(expr$exp, expr$geno, median)
median

```

>Q14: Generate a boxplot with a box per genotype, what could you infer from the relative 
expression value between A/A and G/G displayed in this plot? Does the SNP effect the 
expression of ORMDL3?

From this boxplot, we can infer that having a G/G genotype results in decreased expression compared to a A/A genotype. Therefore, the SNP does effect the expression of ORMDL3, since a change in the nucleotide from an A to a G results in overall decreased expression of ORMDL3. 

Let's make a boxplot of this data:

```{r}
library(ggplot2)

bp <- ggplot(expr) + aes(x=geno, y=exp, fill=geno) + geom_boxplot(notch=TRUE)
bp
```














